### PR TITLE
Remember authentication details with endpoint configuration

### LINKED ISSUE #1067: Remember authentication details with endpoint configuration
__Is your feature request related to a problem? Please describe.__

As a user I am frustrated if I need to re-configure the authentication details every time I re-deploy a diagram.

__Describe the solution you'd like__

* [ ] Deployment details, including authentication type (+ user name) (+ credentials _?_) are stored within the user data `config.json` upon successful deployment and restored on deploy dialog re-open.

__Describe alternatives you've considered__

Don't store the details (password / token) if we don't find a way to secure it.

### PR DESCRIPTION
**Pull request by barmac**
Closes #1067

Add tests:

* [x] fix camunda API tests
* [x] fix deployment tool tests
* [x] fix deployment modal tests
* [x] save flow
* [x] deployment with configuration saved
* [x] deployment with partial configuration
* [x] deployment with no configuration
* [x] endpoint changed => single endpoint right now
* [x] cleanup commit history

### CODE REVIEW

**Comment by philippfromme** [comment]
What's the status of this one? Can I review this?

**Comment by barmac** [comment]
> What's the status of this one? Can I review this?

Yes.

**Comment by barmac** [comment]
@nikku Please test it on your machine. This might need `libsecret` if you don't have it.

**Comment by philippfromme** [comment]
I'm having difficulties running it on Windows.

**Comment by nikku** [comment]
We decided to not ship this with `v3.4` of the app. Will be made available with the next version of the tool.

**Comment by nikku** [comment]
We may follow up on this PR in the future to add OS keychain support.

**Comment by barmac** [comment]
Sure, we can do that. I'll keep the `node-keytar` solution in a separate branch, but implement the standard config-based solution here.

**Comment by barmac** [comment]
@nikku I'd consider this ready for review.

**Comment by nikku** [comment]
I'll have a :eyes:.

File: client/src/app/primitives/DropdownButton.js [hunk 1]
```diff
@@ -8,174 +8,206 @@
...
+    const { disabled } = this.props;
+
+    if (disabled) {
+      return;
+    }
+
+    const { active } = this.state;
+
+    this.setState({
+      active: disabled ||!active
```
Comment (barmac) [inline client/src/app/primitives/DropdownButton.js]:
```suggestion
      active: disabled || !active
```

File: client/src/plugins/deployment-tool/CamundaAPI.js [hunk 2]
```diff
@@ -149,3 +180,114 @@ export default class CamundaAPI {
...
+
+  const PARSE_ERROR_PREFIX = 'ENGINE-09005 Could not parse BPMN process.';
+
+  if (body && body.message && body.message.startsWith(PARSE_ERROR_PREFIX)) {
+    return DIAGRAM_PARSE_ERROR;
+  }
+}
+
+function isLocalhost(url) {
+  return /^https?:\/\/(127\.0\.0\.1|localhost)/.test(url);
```
Comment (barmac) [inline client/src/plugins/deployment-tool/CamundaAPI.js]: Good catch.

File: client/src/plugins/deployment-tool/just-form.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+  } = props;
+
+  let filterValue;
+
+  if (type === 'checkbox') {
+    filterValue = (value) => {
+      console.log(value);
+
+      return value;
+    };
```
Comment (barmac) [inline client/src/plugins/deployment-tool/just-form.js]:
```suggestion
    filterValue = value => value;
```

Comment (barmac) [reply client/src/plugins/deployment-tool/just-form.js]: It seems that Github can't handle multiline suggestions yet.

Comment (barmac) [reply client/src/plugins/deployment-tool/just-form.js]: It should be about removing the `console.log`.

File: client/src/plugins/deployment-tool/just-form.js [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+  }
+
+  const field = useField(name, { validate, filterValue });
+
+  if (Component) {
+    return <Component field={ field } { ...restProps } />;
+  }
+
+  console.log('INVALID USAGE, expected <Field component={ ComponentType } />');
```
Comment (barmac) [inline client/src/plugins/deployment-tool/just-form.js]:
```suggestion
  throw new Error('INVALID USAGE, expected <Field component={ ComponentType } />');
```

File: client/src/plugins/deployment-tool/DeploymentConfigValidator.js [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+    this.checkStart();
+
+    validator.validateConnection(endpoint).then(connectionError => {
+
+      this.checkCompleted(endpoint, {
+        connectionError
+      });
+
+    }).catch(error => {
+      console.error('connection check failed', error);
```
Comment (barmac) [inline client/src/plugins/deployment-tool/DeploymentConfigValidator.js]: Display error?

File: client/src/plugins/deployment-tool/just-form.js [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+ * Camunda licenses this file to you under the MIT; you may not use this file
+ * except in compliance with the MIT License.
+ */
+
+import React from 'react';
+
+import {
+  useForm,
+  useField
+} from '@nikku/react-form';
```
Comment (barmac) [inline client/src/plugins/deployment-tool/just-form.js]: Replace with 3rd party library.

**Comment by barmac** [comment]
Found a bug in the dropdown: I can't access the configuration anymore after it's saved.

**Comment by barmac** [comment]
I will add test cases. Possible scenarios:

* save flow
* deployment with configuration saved
* deployment with partial configuration
* deployment with no configuration
* endpoint changed => single endpoint right now

**Comment by barmac** [comment]
@nikku I'm leaving the `DeploymentDetailsModalSpec` as it is until you finish work with the UI.

**Comment by nikku** [comment]
Yea, that makes sense. Thanks for your help so far!

File: client/src/plugins/deployment-tool/CamundaAPI.js [hunk 7]
```diff
@@ -8,29 +8,33 @@
...
 
 export default class CamundaAPI {
-  constructor(baseUrl) {
-    this.baseUrl = baseUrl;
+
+  constructor(endpoint) {
+
+    this.baseUrl = normalizeBaseURL(endpoint.url);
+
+    this.authentication = this.getAuthentication(endpoint);
```
Comment (pinussilvestrus) [inline client/src/plugins/deployment-tool/CamundaAPI.js]: Really like this one, makes absolute sense 👍

File: client/src/plugins/deployment-tool/DeploymentDetailsModal.js [hunk 8]
```diff
@@ -10,208 +10,191 @@
...
+      isValid: !hasKeys(endpointErrors) && !connectionError,
+      endpointErrors,
+      connectionError
+    });
   }
 
-  componentWillUnmount() {
-    this.mounted = false;
+  componentDidMount() {
+    this.connectionChecker.subscribe({
```
Comment (pinussilvestrus) [inline client/src/plugins/deployment-tool/DeploymentDetailsModal.js]: Cool stuff, this makes the modal so much easier.

File: client/src/plugins/deployment-tool/DeploymentTool.js [hunk 9]
```diff
@@ -90,199 +82,258 @@ export default class DeploymentTool extends PureComponent {
...
       // (2.3.1) Handle user cancelation
-      if (!details) {
+      if (action === 'cancel') {
         return;
       }
 
-      await this.saveDetails(tab, details);
+      configuration = await this.saveConfiguration(tab, userConfiguration);
+
+      if (action === 'save') {
```
Comment (pinussilvestrus) [inline client/src/plugins/deployment-tool/DeploymentTool.js]: Should be ok to keep it internally for future reference.

**Pull request merged**